### ISSUE TITLE
Feature Request: Add read/writable directory outside of public_html for domains

### ISSUE THREAD

**Issue by animeai** [hestiacp/hestiacp#705]
There are several scripts that either require or recommend having certain directories of code stored outside of the web root. Currently the web root for domains is /home/username/web/domain.com/public_html. Scripts such as moodle require a data directory that is not directly accessible from the web but can be read from and written to by apache/nginx. With the current set ups you can add and chown a directory, but then require a custom package for the user due to php open_basedir restrictions. I would recommend adding one (or more) of the following:

- Directory available for all users such as: /home/username/web/domain/data/ 
- Option in HestiaCP hosting panel to add a custom directory within /home/username that can be read/written to by apache/nginx
- Option in HestaCP to create a custom package that adds one of the above to select users

Considerations:
- The directory will need to be included in default HestiaCP backups
- The directory will need to be added to the php open_basedir restricted list to allow reading and writing when open_basedir is enabled
- chmod/chown should be the same as public_html
- The directory will only be accessible via scripts calling specific files and will not be accessible from the internet

**Comment by ScIT-Raphael** [comment]
Thanks for the detailed feature request, currently we do not have enough man power to fullfill all wishes, even when they are good :). If you would like to contribute the features on your own, you can always do a pull request. If you would like to contribute to hestia and talk with the devs, please send us a mail to info@hestiacp.com.

**Comment by FalzoMAD** [comment]
@animeai as far as I remember the /home/%user%/web/%domain%/private folder was intentionally created for that usage (back in vesta already). there were some different templates for apache to switch between using restrictive basedir or not.

however over time a lot has changed especially to the usage of these templates. and with most of the things now moving away from using mod-php towards php-fpm ,the apache templates lost their meaning anyway. 

for the fpm-pool templates the open-basedir settings need to be reconsidered and based on your input I agree that either the path to that private dir should be included or the basedir restriction could be lifted up one level to the domain-folder ...

TL;DR; +1

**Comment by johnnyq** [comment]
I agree, in the current version 1.0.6 you can use the hosting apache / nginx template to bypass restrictive basedir. Its set by default but if you goto the web interface and choose edit under the web domain you can select php and apache template.

**Comment by ifaist0s** [comment]
This is indeed a nice idea.
I just checked [v-backup-user](https://github.com/hestiacp/hestiacp/blob/master/bin/v-backup-user) and I can see in lines [214](https://github.com/hestiacp/hestiacp/blob/a034e4b49d43c0f5736d3077f7fb1277e00d364c/bin/v-backup-user#L214) and [238](https://github.com/hestiacp/hestiacp/blob/a034e4b49d43c0f5736d3077f7fb1277e00d364c/bin/v-backup-user#L238) that the special directory is going to be included in the backup, **IF** it is located (after manual creation via SSH) just above public_html (i.e. inside /home/[USER]/web/[FQDN]/ ). So I suppose no coding is required for the backup part.

As for changing **open_basedir**, I too have that need (for other reasons) and resolve to manually editing the templates with something like: `sed -i 's|open_basedir.*|&:[path_that_needs_to_be_added]|' /usr/local/hestia/data/templates/web/apache2/PHP-*.sh` But I understand this is quite troublesome for a lot of people, so a WebUI feature would help a lot. Too bad I'm not good at making changes to the WebUI and probably I can't help with that :(

**Comment by Lupul** [comment]
Is this similiar with the way /public subfolder is used in laravel/symfony?

**Comment by vvcares** [comment]
> Is this similiar with the way /public subfolder is used in laravel/symfony?

Hi Lupul. Yes, its similar that way.
Even Opencart also needs the same Ref: https://forum.opencart.com/viewtopic.php?t=200701

**Comment by darkpoetry** [comment]
The crux of the problem is that there is no folder (unless you count public_shtml) that isn't exposed included in open_basedir. The "private" folder definitely needs to be included in that like it used to be in Vesta as I recall. That's a pretty established expectation for folks who will be migrating here I think, such as putting your restricted files there so there is no chance they will be available to the public without first passing through a script. 

Attempted a proposed fix:
https://github.com/hestiacp/hestiacp/pull/1406/

**Comment by jaapmarcus** [comment]
Issue has been fixed. Merged the pull request from @darkpoetry  thanks for the commit ❤️
